### ISSUE TITLE
Alignment is limited to 32 bytes

### ISSUE THREAD

**Issue by isuruf** [isuruf/flang#34]
I tried compiling the reference lapack. Library built fine with some cmake changes for identifying flang.
Building the test suite fails with,
```
[  4%] Building Fortran object BLAS/TESTING/CMakeFiles/xblat1c.dir/cblat1.f.obj
flang.exe: warning: argument unused during compilation: '-fbounds-check' [-Wunused-command-line-argument]
warning: overriding the module target triple with x86_64-pc-windows-msvc19.0.24215 [-Woverride-module]
fatal error: error in backend: alignment is limited to 32-bytes
flang.exe: error: clang frontend command failed with exit code 70 (use -v to see invocation)
clang version 4.0.0
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Users\idf2\miniconda3\Library\bin
flang.exe: note: diagnostic msg: PLEASE submit a bug report to  and include the crash backtrace, preprocessed source, and associated run script.
flang.exe: note: diagnostic msg: Error generating preprocessed source(s) - no preprocessable inputs.
NMAKE : fatal error U1077: 'C:\Users\idf2\miniconda3\Library\bin\flang.exe' : return code '0x46'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\nmake.exe"' : return code '0x2'
Stop.
```

**Comment by ghost** [comment]
Need the crash backtrace. Use `-v`.

**Comment by isuruf** [comment]
```
(root) C:\Users\idf2\Lapack\build\BLAS\TESTING>C:\Users\idf2\miniconda3\Library\bin\flang.exe -v  -c C:\Users\idf2\Lapack\BLAS\TESTING\cblat1.f -o CMakeFiles\xblat1c.dir\cblat1.f.obj
clang version 4.0.0
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Users\idf2\miniconda3\Library\bin
 "C:\\Users\\idf2\\miniconda3\\Library\\bin\\flang1" "C:\\Users\\idf2\\Lapack\\BLAS\\TESTING\\cblat1.f" -opt 0 -terse 1 -inform warn -nohpf -nostatic -y 129 2 -inform warn -x 19 0x400000 -quad -x 59 4 -x 15 2 -x 49 0x400004 -x 51 0x20 -x 57 0x4c -x 58 0x10000 -x 124 0x1000 -tp px -x 57 0xfb0000 -x 58 0x78031040 -x 47 0x08 -x 48 4608 -x 49 0x100 -def __LONG_MAX__=2147483647L -def __SIZE_TYPE__=unsigned long long int -def __PTRDIFF_TYPE__=long long int -def _WIN32 -def WIN32 -def _WIN64 -def WIN64 -def __NO_MATH_INLINES -def __x86_64 -def __x86_64__ -def __THROW= -def __extension__= -def __amd_64__amd64__ -def __k8 -def __k8__ -def __PGLLVM__ -nofreeform -vect 48 -y 54 1 -x 70 0x40000000 -y 163 0xc0000000 -x 189 0x10 -stbfile cblat1-a8991b.stb -modexport cblat1-a8991b.cmod -modindex cblat1-a8991b.cmdx -output cblat1-a8991b.ilm
 "C:\\Users\\idf2\\miniconda3\\Library\\bin\\flang2" cblat1-a8991b.ilm -x 6 0x100 -x 42 0x400000 -y 129 4 -x 129 0x400 -y 129 2 -ieee 1 -fn "C:\\Users\\idf2\\Lapack\\BLAS\\TESTING\\cblat1.f" -opt 0 -terse 1 -inform warn -y 129 2 -inform warn -x 51 0x20 -x 119 0xa10000 -x 122 0x40 -x 123 0x1000 -x 127 4 -x 127 17 -x 19 0x400000 -x 28 0x40000 -x 120 0x10000000 -x 70 0x8000 -x 122 1 -x 125 0x20000 -quad -x 59 4 -tp px -x 120 0x1000 -x 124 0x1400 -y 15 2 -x 57 0x3b0000 -x 58 0x48000000 -x 49 0x100 -astype 0 -x 183 4 -x 121 0x800 -x 54 0x10 -x 70 0x40000000 -x 249 40 -x 124 1 -y 163 0xc0000000 -x 189 0x10 -y 189 0x4000000 -x 183 0x10 -stbfile cblat1-a8991b.stb -asm "C:\\Users\\idf2\\AppData\\Local\\Temp\\cblat1-a8991b.ll"
 "C:\\Users\\idf2\\miniconda3\\Library\\bin\\flang.exe" -cc1 -triple x86_64-pc-windows-msvc19.0.24215 -emit-obj -mrelax-all -mincremental-linker-compatible -disable-free -disable-llvm-verifier -discard-value-names -main-file-name cblat1.f -mrelocation-model pic -pic-level 2 -mthread-model posix -fmath-errno -masm-verbose -mconstructor-aliases -munwind-tables -target-cpu x86-64 -momit-leaf-frame-pointer -v -dwarf-column-info -debugger-tuning=gdb -coverage-notes-file "C:\\Users\\idf2\\Lapack\\build\\BLAS\\TESTING\\CMakeFiles\\xblat1c.dir\\cblat1.f.gcno" -resource-dir "C:\\Users\\idf2\\miniconda3\\Library\\bin\\..\\lib\\clang\\4.0.0" -fdebug-compilation-dir "C:\\Users\\idf2\\Lapack\\build\\BLAS\\TESTING" -ferror-limit 19 -fmessage-length 120 -fms-extensions -fms-compatibility -fms-compatibility-version=19.0.24215 -fdelayed-template-parsing -fobjc-runtime=gcc -fdiagnostics-show-option -fcolor-diagnostics -o "CMakeFiles\\xblat1c.dir\\cblat1.f.obj" -x ir "C:\\Users\\idf2\\AppData\\Local\\Temp\\cblat1-a8991b.ll"
clang -cc1 version 4.0.0 based upon LLVM 4.0.0 default target x86_64-pc-windows-msvc
warning: overriding the module target triple with x86_64-pc-windows-msvc19.0.24215 [-Woverride-module]
fatal error: error in backend: alignment is limited to 32-bytes
flang.exe: error: clang frontend command failed with exit code 70 (use -v to see invocation)
clang version 4.0.0
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Users\idf2\miniconda3\Library\bin
flang.exe: note: diagnostic msg: PLEASE submit a bug report to  and include the crash backtrace, preprocessed source, and associated run script.
flang.exe: note: diagnostic msg: Error generating preprocessed source(s) - no preprocessable inputs.
```

**Comment by ghost** [comment]
Run this (are the tempfiles preserved?):

```
 "C:\\Users\\idf2\\miniconda3\\Library\\bin\\flang.exe" -cc1 -triple x86_64-pc-windows-msvc19.0.24215 -emit-obj -mrelax-all -mincremental-linker-compatible -disable-free -disable-llvm-verifier -discard-value-names -main-file-name cblat1.f -mrelocation-model pic -pic-level 2 -mthread-model posix -fmath-errno -masm-verbose -mconstructor-aliases -munwind-tables -target-cpu x86-64 -momit-leaf-frame-pointer -v -dwarf-column-info -debugger-tuning=gdb -coverage-notes-file "C:\\Users\\idf2\\Lapack\\build\\BLAS\\TESTING\\CMakeFiles\\xblat1c.dir\\cblat1.f.gcno" -resource-dir "C:\\Users\\idf2\\miniconda3\\Library\\bin\\..\\lib\\clang\\4.0.0" -fdebug-compilation-dir "C:\\Users\\idf2\\Lapack\\build\\BLAS\\TESTING" -ferror-limit 19 -fmessage-length 120 -fms-extensions -fms-compatibility -fms-compatibility-version=19.0.24215 -fdelayed-template-parsing -fobjc-runtime=gcc -fdiagnostics-show-option -fcolor-diagnostics -o "CMakeFiles\\xblat1c.dir\\cblat1.f.obj" -x ir "C:\\Users\\idf2\\AppData\\Local\\Temp\\cblat1-a8991b.ll"
```

**Comment by isuruf** [comment]
They are not, so I ran the commands manually. It only gives the above error

**Comment by ghost** [comment]
I asked sscalpone (slack channel). If he doesn't answer then we may need to attach x64dbg to the executables. The message is in there somewhere so theoretically we can grep for it.

**Comment by ghost** [comment]
I don't have time to do that now; other options are more time-efficient. You could spend hours in a debugger.

**Comment by ghost** [comment]
Bingo: https://github.com/llvm-mirror/llvm/blob/806f68bbab5118c0252fdeaf1d84fc17ba9dae5c/lib/MC/MCWinCOFFStreamer.cpp#L230

**Comment by ghost** [comment]
Most helpful commit: https://github.com/llvm-mirror/llvm/commit/e4d89ec8de33c9ed3554dcc2b2391b7698dba1e3

**Comment by ghost** [comment]
@isuruf Can you try building with `fpic`?

**Comment by isuruf** [comment]
Isn't `fpic` only for unixes?

**Comment by ghost** [comment]
That's what I thought but try passing `-fpic` to flang and see what happens.

**Comment by isuruf** [comment]
`flang.exe: error: unsupported option '-fpic' for target 'x86_64-pc-windows-msvc'`

**Comment by ghost** [comment]
There goes that idea.

**Comment by ghost** [comment]
Also hit this error here:

https://ci.appveyor.com/project/xoviat/scipy#L56

**Comment by ghost** [comment]
SO question: https://stackoverflow.com/questions/47008306/llvm-fatal-error-error-in-backend-alignment-is-limited-to-32-bytes

**Comment by isuruf** [comment]
Did you compile OpenBLAS with Flang?

**Comment by ghost** [comment]
No, OpenBLAS is from vcpkg.

**Comment by isuruf** [comment]
How is that compiled? MinGW?

**Comment by ghost** [comment]
No ASM + f2c LAPACK.

**Comment by ghost** [comment]
Problem is that scipy will raise the LAPACK version and then no more f2c. Flang will come to the rescue.

**Comment by isuruf** [comment]
Ah okay. Btw, https://github.com/xianyi/OpenBLAS/pull/1256 supports ASM with clang, so that's supported too.

**Comment by ghost** [comment]
I was aware of that (though not aware it was your work). Problem is vcpkg doesn't support clang (though maybe we could ask for it?).

**Comment by ghost** [comment]
@isuruf  Can you test this file?

```
      SUBROUTINE XERROR(MESS,NMESS,L1,L2)
C
C     THIS IS A DUMMY XERROR ROUTINE TO PRINT ERROR MESSAGES WITH NMESS
C     CHARACTERS. L1 AND L2 ARE DUMMY PARAMETERS TO MAKE THIS CALL
C     COMPATIBLE WITH THE SLATEC XERROR ROUTINE. THIS IS A FORTRAN 77
C     ROUTINE.
C
      RETURN
      END
```

**Comment by ghost** [comment]
Actually I think I've isolated this problem.

**Comment by ghost** [comment]
nvmd. Ignore that.

**Comment by ghost** [comment]
Bingo:

```
      DOUBLE PRECISION FUNCTION D1MACH(I)
      COMMON /D9MACH/ CRAY1
      RETURN
      END
```

**Comment by ghost** [comment]
@isuruf The error occurs when there is a `COMMON` statement. What does that mean?

**Comment by ghost** [comment]
Looks like this may be related to flang-compiler/flang#114 except the compiler detected the error on windows vs a runtime crash.

**Comment by isuruf** [comment]
Great. I can confirm that the above source gives the crash.

> @isuruf The error occurs when there is a COMMON statement. What does that mean?

No idea. I don't code in Fortran except in rare cases where I modify legacy code.

**Comment by ghost** [comment]
For reference, the LLVM:

```
; ModuleID = 'test.f'
target datalayout = "e-p:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-pc-win32"
define double @d1mach_(i64* %i) !dbg !17 {
L.entry:
	%d1mach_299 = alloca double, align 8

	br label %L.LB1_305
L.LB1_305:
	br label %L.LB1_302, !dbg !21
L.LB1_302:
	%0 = load double, double* %d1mach_299, align 8, !dbg !22
	ret double %0, !dbg !22
}

; Named metadata
!llvm.module.flags = !{ !1, !2 }
!llvm.dbg.cu = !{ !10 }

; Metadata
!1 = !{ i32 2, !"Dwarf Version", i32 4 }
!2 = !{ i32 1, !"Debug Info Version", i32 3 }
!3 = !DIFile(filename: "test.f", directory: "C:\5cUsers\5cUser\5cDownloads")
; !4 = !DIFile(tag: DW_TAG_file_type, pair: !3)
!4 = !{ i32 41, !3 }
!5 = !{  }
!6 = !{  }
!7 = !{ !17 }
!8 = !{  }
!9 = !{  }
!10 = distinct !DICompileUnit(file: !3, language: DW_LANG_Fortran90, producer: " F90 Flang - 1.5 2017-05-01", enums: !5, retainedTypes: !6, globals: !8, emissionKind: FullDebug, imports: !9)
!11 = !DIBasicType(tag: DW_TAG_base_type, name: "double precision", size: 64, align: 64, encoding: DW_ATE_float)
!12 = !DIDerivedType(tag: DW_TAG_pointer_type, size: 64, align: 64, baseType: !11)
!13 = !DIBasicType(tag: DW_TAG_base_type, name: "integer", size: 32, align: 32, encoding: DW_ATE_signed)
!14 = !{ !12, !13 }
!15 = !DISubroutineType(types: !14)
!16 = !{  }
!17 = distinct !DISubprogram(file: !3, scope: !10, name: "d1mach", line: 1, type: !15, isDefinition: true, unit: !10, variables: !16, scopeLine: 1)
!18 = !DILocation(scope: !17)
!19 = !DILexicalBlock(file: !3, scope: !17, line: 1, column: 1)
!20 = !DILocation(scope: !19)
!21 = !DILocation(line: 3, column: 1, scope: !19)
!22 = !DILocation(line: 4, column: 1, scope: !19)

%structd9mach_ = type < { [4 x i8]  } > 
@d9mach_ = common global %structd9mach_  zeroinitializer, align 64
```

**Comment by isuruf** [comment]
The last line is the issue

**Comment by ghost** [comment]
What is the function of `align 64`? Would `align 32` have any effect?

**Comment by ghost** [comment]
Root cause may be here: https://github.com/flang-compiler/flang/blob/78e3e1d3288ec9b42efd2388bd6e804aab836c86/tools/flang2/flang2exe/cgmain.c#L9651

Try the PR and see what happens.

**Comment by isuruf** [comment]
Ill look at it later this evening. Btw `align 64` is huge. Maybe they meant 64bits instead of 64 bytes.

**Comment by isuruf** [comment]
Fixed in #36
